File: rusturing/lex.rs
use std::fmt;
use std::collections::HashMap;

#[derive(Debug)]
pub enum Token {
    Eof,
    Error,
    ChrImm,
    Ident,
    Movel,
    Mover,
    Print,
    Lcurly,
    Rcurly,
    Lpar,
    Rpar,
    Lbracket,
    Rbracket,
    Comma,
    ChrVar,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

pub struct Lexer {
    s: String,
    off: usize,
    done: bool,
    line: usize,
    line_start: usize,

    token: Token,
    tstr: &str,
}

impl Lexer {
    pub fn new(input: String) -> Lexer {
        Lexer {
            s = input,
            off = 0,
            done = false,
            line = 1,
            line_start = 0,
            token = Token::Error,
            tstr = s[0..0],
        }
    }

    pub fn toktype(&self) -> Token {
        self.token
    }

    pub fn tokstring(&self) -> &str {
        self.tstr
    }

    pub fn expect(&mut self, toktype: Token) -> Result<(), String> {
        if toktype != self.token {
            Err(format!("expected {} but got {}", toktype, self.token))
        } else {
            self.lex();
            Ok(())
        }
    }

    pub fn lex(&mut self) {
        static single_chars: HashMap<char, Token> = HashMap::from([
            ('<', Token::Movel), ('>', Token::mover),
            ('{', Token::Lcurly), ('}', Token::Rcurly),
            ('(', Token::Lpar), (')', Token::Rpar),
            ('[', Token::Lbracket), (']', Token::Rbracket),
            ('=', Token::Print), (',', Token::Comma), ('$', Token::ChrVar)
        ]);

        if self.done {
            return;
        }

        loop {
           // TODO string indexing or iterator? 
        }
    }
}

File: src/abstract/lex.cpp
#include <cctype>
#include <stdexcept>
#include <map>

#include "lex.hpp"

using std::map;
using std::string;
using std::runtime_error;

namespace lex {

string Lexer::substring() const {
    if (type == eof) {
        return "";
    }
    return s.substr(offset, len);
}

void Lexer::lex() {
    if (done)
        return;

    while(isspace(s[cur]))
        ++cur;

    static map<char, TokenType> single_chars = {
        {'<', movel}, {'>', mover},
        {'{', lcurly}, {'}', rcurly},
        {'(', lpar}, {')', rpar},
        {'[', lbracket}, {']', rbracket},
        {'=', print}, {',', comma}, {'$', chr_var}
    };

    char c = s[cur];
    auto single_found = single_chars.find(c);
    if (c == '\0') {
        done = true;
        settoken(eof, 1, cur);
        return;
    } else if (single_found != single_chars.end()) {
        settoken(single_found->second, 1, cur);
        ++cur;
        return;
    } else if (c == '\'') {
        char c1 = s[cur + 1];
        if (c1 == 'x') {
            char c2 = s[cur + 2];
            if (c2 == '\'') {
                settoken(chr_imm, 1, cur + 1);
                cur += 3;
                return;
            } else if (isxdigit(c2)) {
                char c3 = s[cur + 3];
                if (isxdigit(c3)) {
                    char c4 = s[cur + 4];
                    if (c4 == '\'') {
                        settoken(chr_imm, 3, cur + 1);
                        cur += 5;
                        return;
                    }
                }
            }
        } else if (c1 >= ' ' && c1 <= '~') {
            char c2 = s[cur + 2];
            if (c2 == '\'') {
                settoken(chr_imm, 1, cur + 1);
                cur += 3;
                return;
            }
        }
    } else if (isalpha(c) || c == '_') {
        auto start = cur;
        ++cur;
        while (isalnum(s[cur]) || s[cur] == '_')
            ++cur;
        settoken(ident, cur - start, start);
        return;
    }

    done = true;
    settoken(error, 1, cur);
}

string Lexer::tokname(TokenType t) {
    static string names[] = {
        "eof", "error", "chr_imm", "ident",
        "movel", "mover", "print", "lcurly", "rcurly",
        "lpar", "rpar", "lbracket", "rbracket",
        "comma", "chr_var"
    };
    return names[t];
}

void Lexer::expect(TokenType t) {
    if (type != t)
        throw runtime_error("expected " + tokname(t) + " but got " + tokname(type));
    lex();
}

}
